poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+   initialized to zero.
+   ```
+   state = [0; T]
+   state[RATE] = domain_iv
+   ```
+
+2. Return the new sponge:
+   ```
+   PoseidonSponge(
+     PoseidonPermutation,
+     state,
+     mode = absorbing,
+     cache = [],
+     duplex_padding,
+   )
+   ```
+
+> TODO: Should `RATE` be passed directly to `PoseidonSponge.Initialize` instead of being
+> an implicit parameter?
+
+### Internal helper functions
+
+`ZeroExtend` is a function that takes an input of between zero and `RATE` field elements.
+It appends the array `[0; RATE - len(input)]` and returns the result.
+
+`PoseidonSponge.PerformDuplex` takes an input of between zero and `RATE` field elements.
+It proceeds as follows:
+
+1. Apply the domain's padding function to the input, resulting in a padded input of `RATE`
+   field elements.
+
+2. Add the padded input element-wise to the rate portion of `state`:
+   ```
+   for i in [0, RATE):
+       state[i] += padded_input[i]
+   ```
+
+3. Call `PoseidonPermutation(state)`.
+
+4. Return the first `RATE` field elements of `state` as output.
+
+### Absorb
+
+`PoseidonSponge.Absorb` takes a single field element `input`, and proceeds as follows:
+
+- If `mode == absorbing`:
+  - If `len(cache) == RATE`:
+    1. Call `PoseidonSponge.PerformDuplex(cache)`, and drop the result.
+    2. Set `cache` to `[input; 1]`.
+
+  - Else (there is room in `cache`):
+    1. Append `input` to the end of `cache`.
+
+- Else (`mode == squeezing`):
+  1. Drop any remaining cached output elements (by setting `cache` to `[]`).
+  2. Set `mode` to `absorbing`.
+  3. Set `cache` to `[input; 1]`.
+
+### Squeeze
+
+`PoseidonSponge.Squeeze` proceeds as follows:
+
+1. If `mode == squeezing` and `len(cache) == 0`:
+   1. Set `mode` to `absorbing`.
+   2. Set `cache` to `[]` (this is already the case, but for clarity this means there are
+      no elements to be absorbed).
+
+2. If `mode == absorbing`:
+   1. `new_output_elements = PoseidonSponge.PerformDuplex(cache)`.
+   2. Set `mode` to `squeezing`.
+   3. Set `cache` to `new_output_elements`.
+
+3. Remove the first element from `cache` and return it.
+
+## Hash functions
+
+### Merkle tree digest
+
+Merkle trees are built around `arity`-to-1 compression functions, that combine a set of
+nodes from one tree layer into a single node in the next layer.
+
+> TODO: Explain more clearly why the Merkle tree digest outputs `C` elements (as it's
+> confusing on its face that it becomes an `arity`-to-`C` compression function).
+
+`PoseidonHash.MerkleTree` takes as input:
+
+- The `arity` of the Merkle tree, of maximum value 32.
+  - The `RATE` of the sponge is set to `arity`.
+- A `PoseidonPermutation` with the desired properties; in particular, its width `T` must
+  equal `arity + C`, where `C` depends on the desired security level `M`.
+- `nodes`, an array of `arity` field elements.
+
+The digest proceeds as follows:
+
+```
+sponge = PoseidonSponge.Initialize(
+    PoseidonPermutation,
+    (1 << arity) - 1,
```

Comment:
I haven't specified the "Merkle Tree (some leafs may be empty)" variant from section 4.2 of the Poseidon paper, because it is too ambiguous (it does not say how empty leaves should be represented: omitted from the input which is then zero-extended? left as zeroes at their corresponding input positions?) and I don't know of anyone actually using this case.